### PR TITLE
Linkify commit messages and commit bodies

### PR DESCRIPTION
**Pull request by umpox**
## Description

This PR:
- Transforms URLs in commit messages into anchor tags
- Transforms URLs in commit bodies into anchor tags

image

## Test plan

- Tested locally, e.g. (view the content of my first commit on Sourcegraph: https://sourcegraph.test:3443/github.com/sourcegraph/sourcegraph/-/commit/51b68b09a513c57626ebc1b350f3d98fb865eccc)

## App preview:

- [Web](https://sg-web-tr-commit-msg-urls.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (+0.05 kb) | 0.12% (+17.65 kb) 🔺 | 0.15% (+17.60 kb) 🔺 | 0.13% (+1) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/tr/commit-msg-urls/compare-report.html;tab=live_object) for a full comparison between the commits 6fbc966 and ccf9b20 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/web/src/repo/linkifiy/get-links.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                type: 'gh-issue' as const,
+            })
+        }
+    }
+
+    return links
+}
+
+const getLinks = (input: string): LinkFromString[] => {
+    const links = linkifyFind(input)
```
Comment (umpox) [inline client/web/src/repo/linkifiy/get-links.ts:56]: Just calling out why it makes sense to use `linkifyFind` https://github.com/sourcegraph/sourcegraph/pull/49841/files#diff-a13b5117278ecc1012556c820fca675523204e73c8eca592cb7f292b8fa8d8b9R14-R20

(matching URLs in surrounding text is not fun)

File: client/web/src/repo/linkifiy/get-links.test.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import { ExternalServiceKind } from '../../graphql-operations'
+
+import { getLinksFromString } from './get-links'
+
+const externalURL: { url: string; serviceKind: ExternalServiceKind | null } = {
+    url: 'https://github.com/sourcegraph/sourcegraph',
+    serviceKind: ExternalServiceKind.GITHUB,
+}
+
+describe('get-links', () => {
```
Comment (philipp-spiess) [inline client/web/src/repo/linkifiy/get-links.test.ts:10]: Nice!

File: client/web/src/repo/linkifiy/get-links.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+interface LinkFromString {
+    start: number
+    end: number
+    href: string
+    value: string
+    type: 'url' | 'gh-issue'
+}
+
+export const getLinksFromString = ({ input, externalURLs }: GetLinksFromStringParams): LinkFromString[] => {
```
Comment (fkling) [inline client/web/src/repo/linkifiy/get-links.ts:90]: Nit: I think every exported function should have a doc comment.

**Review by philipp-spiess — APPROVED** [review]
Nice!

**Review by limitedmage — APPROVED** [review]
Nice!! 

Should we use our Markdown renderer for commit bodies? I'm torn because GitHub allows Markdown in PR descriptions (which usually end up becoming the commit messages) but doesn't render Markdown in commit messages themselves.

**Comment by umpox** [comment]
@limitedmage It's a good point, I considered this too but also felt like it was potentially a bit much just for commit messages? Unsure, we'd have to make some UI changes for this I would imagine. I'll raise this in #code-exploration

image

File: package.json [hunk 4]
```diff
@@ -438,6 +438,7 @@
     "js-yaml": "^4.1.0",
     "jsonc-parser": "^3.0.0",
     "linguist-languages": "^7.14.0",
+    "linkifyjs": "^4.1.0",
```
Comment (vovakulikov) [inline package.json:441]: From the linkifyjs doc 
>Fast, accurate and small footprint (~20kB minified, ~11kB gzipped)

I understand that we need this so I think it's fine to have it but just was surprised that nowadays 20kb minified counts like a small footprint

**Pull request merged**